DOC-820 `mysql_cdc` input (#174)
redpanda-data/rp-connect-docs: modules/components/pages/inputs/mysql_cdc.adoc

- https://learn.microsoft.com/en-us/fabric/real-time-hub/add-source-mysql-database-cdc[Azure MySQL Database (CDC)^]
- https://cloud.google.com/datastream/docs/configure-cloudsql-mysql[Google Cloud SQL for MySQL^]

--
Self-hosted MySQL::
+
--

- https://dev.mysql.com/doc/refman/8.4/en/replication-options-binary-log.html[Binary Logging Options and Variables^]

--
=====

== Choose a replication mode

You can run the `mysql_cdc` input in one of two modes, depending on whether you need a snapshot of existing data.

- Snapshot mode: Redpanda Connect first captures a snapshot of all data in the selected tables and streams the contents before processing changes from the last recorded binlog position.
- Streaming mode: Redpanda Connect skips the snapshot and processes only the most recent data changes, starting from the latest binlog position.

=== Snapshot mode

If you set the <<stream_snapshot,`stream_snapshot` field>> to `true`, Redpanda Connect connects to your MySQL database and does the following to capture a snapshot of all data in the selected tables:

. Executes the `FLUSH TABLES WITH READ LOCK` query to write any outstanding table updates to disk, and locks the tables. 

> . Runs the `START TRANSACTION WITH CONSISTENT SNAPSHOT` to create a new transaction with a consistent view of all data, capturing the state of the database at the moment the transaction started.

Reviewer: ```suggestion
. Runs the `START TRANSACTION WITH CONSISTENT SNAPSHOT` statement to create a new transaction with a consistent view of all data, capturing the state of the database at the moment the transaction started.
```
For consistency. Or "query"?

Author: Updated.

. Reads the current binlog position.
. Runs the `UNLOCK TABLES` statement to release the database.
. Preserves the initial transaction for data integrity.

NOTE: If the pipeline restarts during this process, Redpanda Connect must start the snapshot capture from scratch to store the current binlog position in the <<checkpoint_cache,`checkpoint_cache`>>.

> After the snapshot is taken, the input executes SELECT statements to extract data from the selected tables in two stages, it:

Reviewer: ```suggestion
After the snapshot is taken, the input executes SELECT statements to extract data from the selected tables in two stages:
```
I find the structure a bit confusing, would it be better to remove "it" and do something like the following for the ordered list?:

```
. The input finds the primary keys of a table.
. It selects the data ordered by primary key.
```

Author: Updated.

. Finds the primary keys of a table.
. Selects the data ordered by primary key.

> Finally, the input uses the stored binlog position to catchup with changes that occurred during snapshot processing.

Reviewer: ```suggestion
Finally, the input uses the stored binlog position to catch up with changes that occurred during snapshot processing.
```

Author: Updated.

    checkpoint_limit: 1024
    batching:
      count: 0
      byte_size: 0
      period: "" # No default (optional) 
      check: "" # No default (optional)
      processors: [] # No default (optional)
```
--
======

The `mysql_cdc` input uses MySQL's https://dev.mysql.com/doc/refman/8.0/en/binary-log.html[binary log (`binlog`)^] to capture changes made to a MySQL database in real time and streams them to Redpanda Connect.

Redpanda Connect allows you to specify which <<tables,database tables>> in your source database to receive changes from. There are also <<choose-a-replication-mode,two replication modes>> to choose from.

== Prerequisites

- MySQL version 8 or later
ifdef::env-cloud[]
- Network access from the cluster where your Redpanda Connect pipeline is running to the source database environment. For detailed networking information, including how to set up a VPC peering connection, see xref:networking:index.adoc[Redpanda Cloud Networking].
endif::[]
- A correctly configured MySQL instance with:
  * Binary logging enabled
  * `binlog_format` variable set to `ROW`
  * `binlog_row_image` variable set to `FULL`
  * A database user with the `REPLICATION SLAVE` privilege

> - The mysqldump tool installed on the operating system where your database is hosted

Reviewer 2: This is not needed

Author: Removed these requirements and the associated resources.

=== Configuration resources

[tabs]
=====
Cloud platforms::
+
--

- https://aws.amazon.com/blogs/database/enable-change-data-capture-on-amazon-rds-for-mysql-applications-that-are-using-xa-transactions/[Change data capture on Amazon RDS for MySQL^]
- https://learn.microsoft.com/en-us/fabric/real-time-hub/add-source-mysql-database-cdc[Azure MySQL Database (CDC)^]
- https://cloud.google.com/datastream/docs/configure-cloudsql-mysql[Google Cloud SQL for MySQL^]

--
Self-hosted MySQL::
+
--

> - https://dev.mysql.com/doc/refman/8.4/en/replication-options-binary-log.html[Binary Logging Options and Variables^]

Author: Is it OK to have version specific references for this. I'm not sure how standard these configuration changes are.

Reviewer 2: Yes it's fine 👍

- https://dev.mysql.com/doc/refman/8.4/en/replication-howto-repuser.html[Creating a User for Replication^]

--
=====

== Choose a replication mode

> You can run the `mysql_cdc` input in streaming or snapshot mode.

Reviewer 2: I really feel like this is more "should I take an initial snapshot or not", because in snapshot mode you still get all the streaming functionality, it's just we additionally send the current state of the database before sending changes.

Author: Updated.

In snapshot mode, Redpanda Connect first takes a snapshot of all data in the selected tables and streams the contents before processing changes from the last recorded binlog position. In streaming mode, Redpanda Connect only processes the most recent changes starting from the latest binlog position.

=== Streaming mode

If you set the <<stream_snapshot,`stream_snapshot` field>> to `false`, Redpanda Connect connects to your MySQL database and starts processing data changes from the latest binlog position. If the pipeline restarts, Redpanda Connect resumes processing updates from the last binlog position written to the <<checkpoint_cache,`checkpoint_cache`>>.

=== Snapshot mode

> If you set the <<stream_snapshot,`stream_snapshot` field>> to `true`, Redpanda Connect connects to your MySQL database and does the following to capture a snapshot of all data in the selected tables:

Author: Does this section include the right level of detail?

Reviewer 2: Yes, this is amazing.

> . Runs the `START TRANSACTION WITH CONSISTENT SNAPSHOT` command to gain exclusive use of the database.

Reviewer 2: We don't get exclusive acccess, that sounds scary to a DBA... The Consistent Snapshot makes sure that you get a consistent view of the data during the snapshot.

The order here Vlad had used was also wrong the correct order is:

* FLUSH TABLES WITH READ LOCK
* START TRANSACTION
* read binlog position
* UNLOCK TABLES

Author: Updated.

. Runs the `UNLOCK TABLES` statement to release the database.
. Preserves the initial transaction for data integrity.

NOTE: If the pipeline restarts during this process, Redpanda Connect must start the snapshot capture from scratch to store the current binlog position in the <<checkpoint_cache,`checkpoint_cache`>>.

After the snapshot is taken, the input executes SELECT statements to extract data from the selected tables in two stages, it:

. Finds the primary keys of a table.
. Selects the data ordered by primary key.

Finally, the input uses the stored binlog position to catchup with changes that occurred during snapshot processing.

== Binlog rotation

While the `mysql_cdc` input is streaming changes to Redpanda Connect, your MySQL server may rotate the binlog file. When this occurs, Redpanda Connect flushes the existing message batch and stores the new binlog position so that it can resume processing using the latest offset.

== Monitor the replication process

You can monitor the initial replication of data using the following metrics:

> TODO

Author: Are there any metrics available?

Reviewer 2: Not currently but we should fix that @jrkinley

Author: Removing this section for now.

== Troubleshoot replication failures

> TODO

Reviewer 2: I think we'll need to run into these before we know. Right now I can't think of any, but maybe @jrkinley has some ideas.

> TODO

Author: Removing this for now.

== Data mappings

The following table shows how selected MySQL data types are mapped to data types supported in Redpanda Connect. All other data types are mapped to string values.

|===

> | MySQL data type | Bloblang value

Author: I took these from the code. Please check for accuracy.

Reviewer 2: Additional ones:

* DATETIME and TIMESTAMP turn into blobl timestamps
* TEXT and VARCHAR types turn into strings

| BINARY, VARBINARY, TINYBLOB, BLOB, MEDIUMBLOB, LONGBLOB
| An array of byte values, for example: `[byte1,byte2,byte3]`

| TINYINT, SMALLINT, MEDIUMINT, INT, BIGINT, YEAR
| A 64-bit integer (`Int64`), for example: `123`

| DECIMAL, NUMERIC

> | JSON number types, for example: `1`

Reviewer 2: These can be combined.

Author: Updated.

| FLOAT, DOUBLE
| A 64-bit decimal (`float64)`, for example: `123.1234`

| SET
| An array of strings, for example: ["apple", "banana", "orange"]

| JSON
| A map object of the JSON, for example: `{"red": 1, "blue": 2, "green": 3}`

|===

== Metadata

This input adds the following metadata fields to each message:

- `operation`: The type of database operation that generated the message, such as `read`, `insert`, `update`, `delete`. A `read` operation occurs when a snapshot of the database is processed.
- `table`: The name of the database table from which the message originated.

> - `binlog_position`: The https://dev.mysql.com/doc/refman/8.0/en/binary-log.html[Binary Log (binlog)^] position of each data update from the source MySQL database. The `binlog` values are strings that you can sort to determine the order in which data updates occurred.

Reviewer 2: Snapshot streamed values do not have a binlog position.

Author: Updated.

== Fields

=== `dsn`

The data source name (DSN) of the MySQL database from which you want to stream updates. Use the format `user:password@tcp(localhost:3306)/database`.

*Type*: `string`

```yml
# Examples

dsn: my_username:my_password@tcp(localhost:3306)/my_db
```

=== `tables`

A list of the database table names to stream changes from. Specify each table name as a separate item.

*Type*: `array`

```yml
# Examples

tables:
  - orders_table
  - customer_address_table
  - inventory_table
```

=== `checkpoint_cache`

Specify a `cache` resource to store the binlog position of the most recent data update delivered to Redpanda Connect. After a restart, Redpanda Connect can continue processing changes from this last known position, avoiding the need to reprocess all table updates.

*Type*: `string`

*Default*: `""`

```yml

> # Examples

Author: Could we give an example as this seems fairly fundamental. Maybe something similar to https://docs.redpanda.com/redpanda-connect/components/inputs/spicedb_watch/#configure-a-cache?

Reviewer 2: Yeah agreed, that example seems good.

```
input:
  mysql_cdc:
    dsn: username:password@tcp(localhost:3306)/database
    tables: [my_table]
    checkpoint_cache: "my_cdc_cache"
cache_resources:
  - label: "my_cdc_cache"
    redis:
      url: redis://:6379
```

Author: Thanks for the example! I've added it in.

*Type*: `int`

*Default*: `0`

=== `batching.byte_size`

The number of bytes at which the batch is flushed. Set to `0` to disable size-based batching.

*Type*: `int`

*Default*: `0`

=== `batching.period`

The period of time after which an incomplete batch is flushed regardless of its size.

*Type*: `string`

*Default*: `""`

```yml
# Examples

period: 1s

period: 1m

period: 500ms
```
=== `batching.check`

A xref:guides:bloblang/about.adoc[Bloblang query] that returns a boolean value indicating whether a message should end a batch.

*Type*: `string`

*Default*: `""`

```yml
# Examples

check: this.type == "end_of_transaction"
```

=== `batching.processors`

For aggregating and archiving message batches, you can add a list of xref:components:processors/about.adoc[processors] to apply to a batch as it is flushed. All resulting messages are flushed as a single batch even when you configure processors to split the batch into smaller batches.

*Type*: `array`

```yml
# Examples

processors:
  - archive:
      format: concatenate

processors:
  - archive:
      format: lines

processors:
  - archive:
      format: json_array
```

== Example pipeline

> TODO

Author: Do you have any examples that you have used for demos etc that I can use here?

Reviewer 2: @garrett-raska do you have that mysql direct into snowflake config that you can share here?

Author: Removed placeholder for now until and example is available.

- `operation`: The type of database operation that generated the message, such as `read`, `insert`, `update`, `delete`. A `read` operation occurs when a snapshot of the database is processed.
- `table`: The name of the database table from which the message originated.

> - `binlog_position`: The https://dev.mysql.com/doc/refman/8.0/en/binary-log.html[Binary Log (binlog)^] position of each data update streamed from the source MySQL database, excluding the initial snapshot. The `binlog` values are strings that you can sort to determine the order in which data updates occurred.

Reviewer 2: I wouldn't say "excluding the initial snapshot", but something like: "data from the initial snapshot does not have a binlog_position set"

Author: Updated
